tools/cmake/UploadMethods.md

```diff
@@ -0,0 +1,214 @@
+Mbed OS's CMake build system supports a number of upload methods to help get code onto your device.  Some uploaders also support debugging, and for these CMake will help you by automatically running the needed commands to start a debugging session.
+
+In order to enable uploading, you'll need to configure which upload methods your project supports.  To do this, you'll need to create an UploadMethodConfig.cmake file and include it in your top-level CMakeLists.txt.  You can start by using the template UploadMethodConfig-Sample.cmake (which is for Nucleo F429ZI) in this folder.  You'll then edit it to control the upload methods that are available for your target, and set the parameters that they need to operate.  THe different parameters and their values are explained in this document.
+
+This page lists each upload method, its _parameters_ (variables that will have a fixed value) and its _options_ (variables that are different for different people and should be set on the command line when you run cmake).  Since this config is set by standard CMake code, it's very easy to add your own additional logic (including command line options) to adjust these variables!
+
+Once you configure some upload methods, you can then run cmake with the `-DUPLOAD_METHOD=<method>` argument to enable uploading your code to the target.
+
+To upload an executable, just run `make flash-xxx`, where xxx is replaced by the name of the executable target.  Don't forget that you can also use `make help` to see the list of all available targets.  Also, for Ninja users, just replace `make` with `ninja` in these commands.
+
+To debug, first run `make gdbserver` in one terminal to start up the debug probe server.  Then, you can connect to the GDB server in your debug tool of choice, or run `make debug-xxx` in another terminal to start command-line GDB on the given program.
+
+**All current upload methods (see below for details):**
+|Name|CMake Argument|Supports Uploading|Supports Debugging|Upload Speed|Supported On
+|---|---|---|---|---|---|
+|No upload method|`-DUPLOAD_METHOD=NONE`|❌|❌|N/A|N/A
+|MBed USB|`-DUPLOAD_METHOD=MBED`|✔️|❌|Fast|All Mbed Boards
+|J-Link|`-DUPLOAD_METHOD=JLINK`|✔️|✔️|Fast|Mbed boards with J-Link On-Board.  Custom boards with a J-Link probe.
+|pyOCD|`-DUPLOAD_METHOD=PYOCD`|✔️|✔️|Medium|Almost all Mbed boards.  Custom boards with an ST-Link or DAPLink probe.
+|OpenOCD|`-DUPLOAD_METHOD=OPENOCD`|✔️|✔️|(Varies)|A number of different debug probes, but requires configuration.
+|NXP Bootloader|`-DUPLOAD_METHOD=NXPPROG`|✔️|❌|Slow|LPC1768 and LPC1114 on custom boards.
```

Comment:
I haven't seen this before, is it used anywhere (as most of the boards are actually supported by either openocd/mbed usb).